### DISCUSSION TITLE
Dynamic object validation

### DISCUSSION THREAD

**Discussion by jakubburzynski** [colinhacks/zod#2783] [Q&A] [answered]
Hi,
at first I want to mention that I already spent some time looking for similar problem, but didn't find anything close to it. 
### Context
I'm working on some NestJS app with already installed Zod for requests validation. Currently I'm trying to create validation for custom configuration file.
For the sake of simplicity, here is the example configuration object, even though it's going to be loaded from .yml file.
```ts
const config = {
  languages: ['en', 'de'],
  resources: [
    {
      names: {
        en: 'Apple',
        de: 'Apfel',
      },
    },
    {
      names: {
        en: 'Cherry',
        de: 'Kirsche',
      },
    },
  ],
};
```
The main goal of validation is to require setting every name translation, according to entered languages. For example if I choose English only, then config should look like this:
```ts
const config = {
  languages: ['en'],
  resources: [
    {
      names: {
        en: 'Apple',
      },
    },
    {
      names: {
        en: 'Cherry',
      },
    },
  ],
};
```
If I choose more languages it should look like this:
```ts
const config = {
  languages: ['en', 'de', 'pl'],
  resources: [
    {
      names: {
        en: 'Apple',
        de: 'Apfel',
        pl: 'Jabłko',
      },
    },
    {
      names: {
        en: 'Cherry',
        de: 'Kirsche',
        pl: 'Wiśnia',
      },
    },
  ],
};
```
Of course, every missing or excess translation should result in validation error. So next two examples are considered invalid.
```ts
const config = {
  languages: ['en'], // missing 'de'
  resources: [
    {
      names: {
        en: 'Apple',
        de: 'Apfel',
      },
    },
    {
      names: {
        en: 'Cherry',
        de: 'Kirsche',
      },
    },
  ],
};
```
```ts
const config = {
  languages: ['en', 'de', 'pl'],
  resources: [
    {
      names: {
        en: 'Apple',
        de: 'Apfel',
        // missing pl:
      },
    },
    {
      names: {
        en: 'Cherry',
        de: 'Kirsche',
        pl: 'Wiśnia',
      },
    },
  ],
};
```
As for now I came up with two step solution. At first it checks an array of languages and then uses it to construct a resources schema validation.
```ts
const supportedLanguages = ['en', 'pl', 'de'] as const;
const languagesSchema = z.array(z.enum(supportedLanguages)).nonempty();
const languages = languagesSchema.parse(config.languages);
console.log(languages);

const resourcesSchema = z.array(
  z.object({
    names: z.object(
      languages.reduce(
        (acc, lang) => ({ ...acc, [lang]: z.string() }),
        {} as Partial<Record<(typeof languages)[number], z.ZodString>>,
      ),
    ),
  }),
);
const resources = resourcesSchema.parse(config.resources);
console.log(resources);
```
### Problem
It works well, but the only problem is that resource names are of type `unknown` rather than `string`.
![image](https://github.com/colinhacks/zod/assets/33811303/22812f70-5e71-4554-a89f-7f3647f0ad25)

Any help will be appreciated, even if it doesn't resolve `unknown` type issue, but maybe changes config structure or schema construction.

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const config = {
    languages: [ 'en' ], // missing 'de'
    resources: [
        {
            names: {
                en: 'Apple',
                de: 'Apfel',
            },
        },
        {
            names: {
                en: 'Cherry',
                de: 'Kirsche',
            },
        },
    ],
}

const supportedLanguages = [ 'en', 'pl', 'de' ] as const
const languagesSchema = z.enum( supportedLanguages ).array().nonempty()
const languages = languagesSchema.parse( config.languages )

const resourcesSchema = z.object( {
    names: z.record( z.enum( languages ), z.string() ),
} ).array()
const result = resourcesSchema.safeParse( config.resources )
!result.success && console.log( result.error.issues )
// [
//     {
//         received: "de",
//         code: "invalid_enum_value",
//         options: [ "en" ],
//         path: [ 0, "names", "de" ],
//         message: "Invalid enum value. Expected 'en', received 'de'"
//     }, {
//         received: "de",
//         code: "invalid_enum_value",
//         options: [ "en" ],
//         path: [ 1, "names", "de" ],
//         message: "Invalid enum value. Expected 'en', received 'de'"
//     }
// ]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by jakubburzynski** [reply]
Hi,
I think it resolves the issue. I also came up with 'multi-parse' solution, but did it a little differently and it wasn't as effective and easy as your way.

Thanks for your help.

**Reply by JacobWeisenburger** [reply]
No need to close the discussion, I just marked it as answered and that's good enough. That way other people can find this if needed.
